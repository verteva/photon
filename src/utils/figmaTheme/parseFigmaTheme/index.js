export { parseFigmaJson, flattenObjectToCssVars } from './parseFigmaTheme';
