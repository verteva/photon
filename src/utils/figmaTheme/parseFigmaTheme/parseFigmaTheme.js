import { replaceTemplateValue } from '../templateResolvers';

export const parseFigmaJson = ({ sourceTheme, semanticTheme }) => {
  return parseObject(semanticTheme, sourceTheme);
};

export const parseObject = (object, root) => {
  const parsedObject = {};

  Object.keys(object).forEach((key) => {
    if (typeof object[key] !== 'object') {
      throw new Error(
        'Invalid template. Cannot parse non-object values. Objects must either have nested objects or terminate with an object with a valid type property.'
      );
    }
    // Only run if type-value pair is found
    if (object[key].type && object[key].value) {
      parsedObject[key] = replaceTemplateValue(object[key], root);
      return;
    }

    parsedObject[key] = parseObject(object[key], root);
  });

  return parsedObject;
};

export const flattenObjectToCssVars = (
  object,
  base = '--',
  delimiter = '-'
) => {
  const flattenedObject = {};

  Object.keys(object).forEach((key) => {
    if (typeof object[key] === 'object') {
      Object.assign(
        flattenedObject,
        flattenObjectToCssVars(object[key], base + key + delimiter)
      );
    } else {
      flattenedObject[base + key] = object[key];
    }
  });

  return flattenedObject;
};
