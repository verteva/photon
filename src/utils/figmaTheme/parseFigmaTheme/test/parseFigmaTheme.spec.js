import { parseFigmaJson, flattenObjectToCssVars } from '../parseFigmaTheme';
import rawTheme from './rawTheme.json';
import rawErrorTheme from './rawErrorTheme.json';
import rawSourceTheme from './rawSourceTheme.json';
import rawSemanticTheme from './rawSemanticTheme.json';

describe('figmaTheme/parseFigmaTheme', () => {
  it('parses a figma theme', () => {
    const result = parseFigmaJson({
      sourceTheme: rawTheme,
      semanticTheme: rawTheme,
    });
    expect(result).toMatchSnapshot();
  });

  it('parses a figma theme from a semantic and source file', () => {
    const result = parseFigmaJson({
      sourceTheme: rawSourceTheme,
      semanticTheme: rawSemanticTheme,
    });
    expect(result).toMatchSnapshot();
  });

  describe('errors', () => {
    it('throws when the them file format is wrong', () => {
      expect(() =>
        parseFigmaJson({
          sourceTheme: rawErrorTheme,
          semanticTheme: rawErrorTheme,
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('figmaTheme/flattenObjectToCssVars', () => {
  it('parses a figma theme and then flattens it', () => {
    const result = parseFigmaJson({
      sourceTheme: rawTheme,
      semanticTheme: rawTheme,
    });
    const cssResult = flattenObjectToCssVars(result);
    expect(cssResult).toMatchSnapshot();
  });
});
