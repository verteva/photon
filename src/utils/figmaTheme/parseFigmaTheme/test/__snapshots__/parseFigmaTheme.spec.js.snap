// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`figmaTheme/flattenObjectToCssVars parses a figma theme and then flattens it 1`] = `
Object {
  "--colors-black": "#000000",
  "--colors-brand-green": "#219653",
  "--colors-brand-orange": "orange",
  "--colors-brand-purple": "#6f53f9",
  "--colors-brand-red": "#eb5757",
  "--colors-brand-teal": "#31839b",
  "--colors-grey-00": "#07131f",
  "--colors-grey-01": "#6a6e80",
  "--colors-grey-02": "#949494",
  "--colors-grey-03": "#b5b5b5",
  "--colors-grey-08": "#f4f4f4",
  "--colors-grey-09": "grey",
  "--colors-grey-10": "#f9f9f9",
  "--colors-theme-primary": "",
  "--colors-theme-secondary": "#31839b",
  "--colors-theme-tertiary": "#eb5757",
  "--colors-white": "#ffffff",
  "--spacing-default": "2",
}
`;

exports[`figmaTheme/parseFigmaTheme errors throws when the them file format is wrong 1`] = `"Invalid template. Cannot parse non-object values. Objects must either have nested objects or terminate with an object with a valid type property."`;

exports[`figmaTheme/parseFigmaTheme parses a figma theme 1`] = `
Object {
  "colors": Object {
    "black": "#000000",
    "brand": Object {
      "green": "#219653",
      "orange": "orange",
      "purple": "#6f53f9",
      "red": "#eb5757",
      "teal": "#31839b",
    },
    "grey": Object {
      "00": "#07131f",
      "01": "#6a6e80",
      "02": "#949494",
      "03": "#b5b5b5",
      "08": "#f4f4f4",
      "09": "grey",
      "10": "#f9f9f9",
    },
    "theme": Object {
      "primary": "",
      "secondary": "#31839b",
      "tertiary": "#eb5757",
    },
    "white": "#ffffff",
  },
  "spacing": Object {
    "default": "2",
  },
}
`;
